streamproducer: retry loading a job in the c2c stream producer

lidorcarmel opened the pull request:
See commits.

Epic: none
Fixes: #107989

Release note: None

Commit:
streamproducer: retry loading a job in the c2c stream producer

Previously, a single LoadJob failure caused the c2c producer job to fail,
which can happen when the node is overloaded.

This patch adds retries, so that a producer failure will be less common.

Epic: none
Fixes: #107989

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107991)
<!-- Reviewable:end -->

stevendanna commented on pkg/ccl/streamingccl/streamproducer/producer_job.go:
Lets mention that this came from loading the job

lidorcarmel replied:
done.

stevendanna reviewed (commented):
This is certainly an improvement, but I wonder if we should go further.  Namely, perhaps we should retry for much longer and then moved to a paused state rather than a failed state.

Failing here is pretty drastic as it means that you potentially need to start a whole new tenant replication. 

Thoughts?

lidorcarmel commented:
PTAL.
Thanks for the chat yesterday, this is still only the retry on the producer side, I'll have another PR for the ingestion side and another for changing the expiration to be last heartbeat as we discussed.

stevendanna commented:
This looks good to me still.

Commit:
streamingest: retry heartbeating the source cluster on failure

Previously we failed when the destination cluster could not successfully
get a heartbeat response from the source cluster.

This is not good for resilience of course - heartbeats can fail because
of many reasons and replication should not halt in those cases.

This commit retries on those failures. Additionally, this pr adds some basic
unit tests around that one-line change.

Epic: none
Fixes: #107989

Release note: None

lidorcarmel commented:
hi @stevendanna! finally got back to this, you already reviewed the second commit, but the new test there was flaky because we didn't retry on the destination, so I added the first commit to do that retry. please review that 'retry heartbeating' commit.

stevendanna reviewed (approved):
Really nice testing here.

I think we discussed some strategies for putting an upper bound on this retrying by keeping the last-seen progress in memory and using that to drive some eventual expiration. But, let's keep that for a follow up.

I think we should backport this one.

lidorcarmel reviewed (commented):
thanks Steven!
I think you're traveling so I'll merge, but if anything comes up later I'm happy to do a followup.

lidorcarmel commented:
TFTR! and also thanks for the discussions Steven!
bors r+
